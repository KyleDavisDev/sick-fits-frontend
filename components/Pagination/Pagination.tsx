import * as React from "react";
import { Query } from "react-apollo";
import gql from "graphql-tag";
import Head from "next/head";
import Link from "next/link";

import { perPage } from "../../config";

export const PAGINATION_QUERY = gql`
  query PAGINATION_QUERY {
    itemsConnection {
      aggregate {
        count
      }
    }
  }
`;

import { PaginationStyles } from "./PaginationStyles";

interface IPaginationProps {
  page: number;
}

const Pagination: React.FunctionComponent<IPaginationProps> = props => {
  return (
    <Query query={PAGINATION_QUERY}>
      {({ data, loading }) => {
        const count = data.itemsConnection.aggregate.count;
        const pages = Math.ceil(count / perPage);
        const curPage = props.page;
        if (loading) return <p>Loading...</p>;
        return (
          <React.Fragment>
            <Head>
              <title>
                Sick Fits! Page {curPage} of {pages}
              </title>
            </Head>
            <PaginationStyles>
              <Link href={{ pathname: "shop", query: { page: curPage - 1 } }}>
                <a className="prev" aria-disabled={curPage <= 1}>
                  &larr; Prev
                </a>
              </Link>
              <p>
                Page {curPage} of {pages} !
              </p>
              <Link href={{ pathname: "shop", query: { page: curPage + 1 } }}>
                <a className="next" aria-disabled={curPage >= pages}>
                  Next &rarr;
                </a>
              </Link>
            </PaginationStyles>
          </React.Fragment>
        );
      }}
    </Query>
  );
};

export default Pagination;
