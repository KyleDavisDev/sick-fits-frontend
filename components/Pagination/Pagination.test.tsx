import "jsdom-global/register";
import * as React from "react";
import { mount } from "enzyme";
import wait from "waait";
import { MockedProvider } from "react-apollo/test-utils";
import Pagination, { PAGINATION_QUERY } from "./Pagination";
import { perPage } from "../../config";
let page = 1;

function makeMocksFor(length: number) {
  return [
    {
      request: { query: PAGINATION_QUERY },
      result: {
        data: {
          itemsConnection: {
            __typename: "aggregate",
            aggregate: { __typename: "count", count: length }
          }
        }
      }
    }
  ];
}

describe("<Pagination />", () => {
  it("renders", () => {
    const wrapper = mount(
      <MockedProvider mocks={makeMocksFor(page)}>
        <Pagination page={page} />
      </MockedProvider>
    );

    expect(wrapper).toBeTruthy();
  });

  it("matches snapshot", async () => {
    const wrapper = mount(
      <MockedProvider mocks={makeMocksFor(page)}>
        <Pagination page={page} />
      </MockedProvider>
    );

    expect(wrapper.find("p")).toMatchSnapshot();

    // let component load/update
    await wait();
    wrapper.update();

    // check links
    expect(
      wrapper.find("[data-test='pagination']").find("Link")
    ).toMatchSnapshot();

    // check head
    expect(wrapper.find("Head")).toMatchSnapshot();
  });

  it("renders loading text", () => {
    const wrapper = mount(
      <MockedProvider mocks={makeMocksFor(page)}>
        <Pagination page={page} />
      </MockedProvider>
    );

    expect(wrapper.find("p").text()).toContain("Loading");
  });

  it("renders correct text many items", async () => {
    const ranNum = Math.floor(Math.random() * 100);
    const wrapper = mount(
      <MockedProvider mocks={makeMocksFor(ranNum)}>
        <Pagination page={page} />
      </MockedProvider>
    );

    // let component load/update
    await wait();
    wrapper.update();

    // grab where page count is being displayed
    const pageContainer = wrapper.find("[data-test='pagination'] p");

    // check text
    expect(pageContainer.text()).toContain(
      `Page ${page} of ${Math.ceil(ranNum / perPage)} !`
    );
  });

  it("renders disabled previous button on first page", async () => {
    const ranNum = Math.floor(Math.random() * 100);
    const wrapper = mount(
      <MockedProvider mocks={makeMocksFor(ranNum)}>
        <Pagination page={page} />
      </MockedProvider>
    );

    // let component load/update
    await wait();
    wrapper.update();

    expect(wrapper.find("a.prev").prop("aria-disabled")).toEqual(true);
  });

  it("renders disabled next button on last page", async () => {
    const ranNum = Math.floor(Math.random() * 100);
    const wrapper = mount(
      <MockedProvider mocks={makeMocksFor(ranNum)}>
        <Pagination page={ranNum} />
      </MockedProvider>
    );

    // let component load/update
    await wait();
    wrapper.update();

    expect(wrapper.find("a.next").prop("aria-disabled")).toEqual(true);
  });

  it("renders enabled buttons on middle page", async () => {
    const wrapper = mount(
      <MockedProvider mocks={makeMocksFor(20)}>
        <Pagination page={3} />
      </MockedProvider>
    );

    // let component load/update
    await wait();
    wrapper.update();

    expect(wrapper.find("a.prev").prop("aria-disabled")).toEqual(false);
    expect(wrapper.find("a.next").prop("aria-disabled")).toEqual(false);
  });
});
