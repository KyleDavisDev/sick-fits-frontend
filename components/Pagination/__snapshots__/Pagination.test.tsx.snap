// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> matches snapshot 1`] = `
<p>
  Loading...
</p>
`;

exports[`<Pagination /> matches snapshot 2`] = `
Array [
  <Link
    href={
      Object {
        "pathname": "shop",
        "query": Object {
          "page": 4,
        },
      }
    }
  >
    <a
      aria-disabled={false}
      className="prev"
      href="shop?page=4"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      ← Prev
    </a>
  </Link>,
  <Link
    href={
      Object {
        "pathname": "shop",
        "query": Object {
          "page": 6,
        },
      }
    }
  >
    <a
      aria-disabled={true}
      className="next"
      href="shop?page=6"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      Next →
    </a>
  </Link>,
]
`;

exports[`<Pagination /> matches snapshot 3`] = `
<Head>
  <Component
    handleStateChange={null}
    inAmpMode={false}
    reduceComponentsToState={[Function]}
  />
</Head>
`;
